app/models/piece.rb

```diff
@@ -20,46 +29,104 @@ def move_to!(new_x, new_y)
     end
     # Finally, it should call update_attributes on the piece and change the piece's x/y position.
     # http://apidock.com/rails/ActiveRecord/Base/update_attributes
-    raise ArgumentError, "Can't move piece" unless update_attributes(x_coordinate: new_x, y_coordinate: new_y)
+    raise ArgumentError, "Can't move piece" unless update_attributes(x_coordinate: x, y_coordinate: y)
   end
 
-  def on_board?(destination_x, destination_y)
-    return false if destination_x < 0
-    return false if destination_x > 7
-    return false if destination_y < 0
-    return false if destination_y > 7
-    true
+  def obstructed?(x, y)
+    # only check if obstructed in the direction you are moving
+    return true if vertical?(x, y) && obstructed_vertically?(y)
+    return true if horizontal?(x, y) && obstructed_horizontally?(x)
+    return true if diagonal?(x, y) && obstructed_diagonally?(x, y)
```

Comment:
Before, if you moved diagonally, it might still say you are obstructed, if there was a piece on the horizontal or vertical. 

for example a queen moving from (0, 0) to (4, 4), might not have anything obstructing the diagonal.  But a piece at (2,0) would still be obstructing it horizontally.  That shouldn't matter if it's only moving diagonally.